File: AlternatePositiveAndNegativeNumbers/AlternatePositiveAndNegativeNumbers.cpp
/*
https://www.geeksforgeeks.org/problems/array-of-alternate-ve-and-ve-nos1401/1
*/

//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
// User function template for C++
class Solution {
  public:

    void rearrange(vector<int> &arr) {
        int sz=arr.size();
        vector<int>positives, negatives;
        for(auto &vals: arr){
            if(vals>=0) positives.push_back(vals);
            else negatives.push_back(vals);
        }
        int idx1=0, idx2=0;
        for(int i=0; i<sz; ++i){
            if(i%2==0){
                if(positives.size()>=1&&idx1<=positives.size()-1){
                    arr[i]=positives[idx1++];
                } else{
                    arr[i]=negatives[idx2++];
                }
            } else{
                if(negatives.size()>=1&&idx2<=negatives.size()-1){
                    arr[i]=negatives[idx2++];
                } else{
                    arr[i]=positives[idx1++];
                }
            }
        }
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        string input;
        int num;

        getline(cin, input);
        stringstream s2(input);
        while (s2 >> num) {
            arr.push_back(num);
        }
        Solution ob;
        ob.rearrange(arr);
        for (int i = 0; i < arr.size(); i++) {
            cout << arr[i] << " ";
        }
        cout << "\n";
    }
    return 0;
}

// } Driver Code Ends

File: AlternatePositiveAndNegativeNumbers/AlternatePositiveAndNegativeNumbers.java
/*
https://www.geeksforgeeks.org/problems/array-of-alternate-ve-and-ve-nos1401/1
*/

//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine().trim());
        while (tc-- > 0) {
            String input = br.readLine();
            String[] inputArray = input.split("\\s+");
            ArrayList<Integer> arr = new ArrayList<>();

            for (String s : inputArray) {
                arr.add(Integer.parseInt(s));
            }

            new Solution().rearrange(arr);
            for (int num : arr) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    void rearrange(ArrayList<Integer> arr) {
        int sizeOfArr=arr.size();
        ArrayList<Integer>positiveNumbers=new ArrayList<>();
        ArrayList<Integer>negativeNumbers=new ArrayList<>();
        for(int i=0; i<sizeOfArr; ++i){
            if(arr.get(i)>=0){
                positiveNumbers.add(arr.get(i));
            } else{
                negativeNumbers.add(arr.get(i));
            }
        }
        int beginIndexForPositives=0, beginIndexForNegatives=0;
        for(int i=0; i<sizeOfArr; ++i){
            if(i%2==0){
                if(positiveNumbers.size()>=1&&beginIndexForPositives<=positiveNumbers.size()-1){
                    arr.set(i, positiveNumbers.get(beginIndexForPositives++));
                } else{
                    arr.set(i, negativeNumbers.get(beginIndexForNegatives++));
                }
            } else{
                if(negativeNumbers.size()>=1&&beginIndexForNegatives<=negativeNumbers.size()-1){
                    arr.set(i, negativeNumbers.get(beginIndexForNegatives++));
                } else{
                    arr.set(i, positiveNumbers.get(beginIndexForPositives++));
                }
            }
        }
    }
}